source/client_channel_handler.c

```diff
@@ -36,21 +47,67 @@ static int s_packet_handler_default(
     return aws_raise_error(AWS_ERROR_MQTT_INVALID_PACKET_TYPE);
 }
 
+static void s_schedule_ping_get_outbound_delta_data(
+    struct aws_mqtt_client_connection *connection,
+    uint64_t *result_now,
+    uint64_t *result_outbound_delta,
+    uint64_t *result_ping_time_as_ns) {
+    uint64_t now = 0;
+    aws_channel_current_clock_time(connection->slot->channel, &now);
+
+    uint64_t outbound_delta = 0;
+    // We just need the delta, so biggest minus smallest. We do NOT want to overflow
+    if (aws_sub_u64_checked(now, connection->last_outbound_socket_write_time, &outbound_delta) != AWS_OP_SUCCESS) {
+        aws_sub_u64_checked(connection->last_outbound_socket_write_time, now, &outbound_delta);
+    }
+
+    uint64_t ping_time_ns =
+        aws_timestamp_convert(connection->keep_alive_time_secs, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL);
+
+    // If there has not been a socket write yet, then the delta should be the scheduled ping timeout.
+    if (connection->last_outbound_socket_write_time == 0) {
```

Comment:
We can also do the zero check before `aws_sub_u64_checked` to optimize a little bit. Also, if we do apply the change, we can also reset `last_outbound_socket_write_time` to 0 in `aws_mqtt_client_connection_connect()`, so that we dont need to calculate the delta time when it comes to a new connection.